### PR TITLE
[BEAM-3072] updates to that the error handling and collected the fail…

### PR DESCRIPTION
**Pull request by NikeNano**
Updates to collect more of the error message. This is done by tracing the error message and collect the output from the subprocess during failure.  In order to do this i have added a dependency on the package traceback. 

First contribution to Apache Beam, so feel free to give as much feedback as possible. 
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [x] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by NikeNano** [comment]
@robinyqiu or  @pabloem cloud you please review. Feel free to give a lot of feedback. First contribution and would love to improve :) Thanks!

**Comment by NikeNano** [comment]
Closed by misstake.  Could you please review the pull request @robinyqiu  or @pabloem. Feel free to leave a lot of feedback, my first contribution and would love to learn as much as possible. Thanks!

**Comment by NikeNano** [comment]
Will check the errors and update.

**Comment by pabloem** [comment]
Hi Niklas! Happy to have your contribution. The PreCommit failed, and as a tip to debug these tests with lots of text, you can go to Details > Console Output, then scroll to the Bottom, and you'll find a link to a Gradle scan like so:
image

In the grade scan you can find exactly which tasks failed (in this case, py27 lint, and py27-gcp). The py27 GCP failure looks unrelated, but the lint failures look related to your change: https://scans.gradle.com/s/4ljrrjerslxiy/console-log?task=:beam-sdks-python:lintPy27

To run these checks before pushing your change you can do `tox -e py27-lint` from within `sdks/python` if you have your virtualenv set up, and tox installed.

In short: Thanks for the contribution. Could you fix the lint issues?

**Comment by aaltay** [comment]
@NikeNano Were you able to test the changes and see that errors are captured as expected?

**Comment by NikeNano** [comment]
@pabloem I will fix it :),  @aaltay as far as i understand the ticket, yes :)

File: sdks/python/scripts/run_pylint.sh [hunk 1]
```diff
@@ -24,6 +24,13 @@
 # The exit-code of the script indicates success or a failure.
 
 # Check that the script is running in a known directory.
+echo "START"
```
Comment (NikeNano) [inline sdks/python/scripts/run_pylint.sh]: my bad will remove this, just wanted to check and try to understand some things....

**Comment by NikeNano** [comment]
@aaltay after thinking about it a bit more, could you elaborate on how "hat errors are captured as expected" and i will add some tests for it. I have set it up so that errors ar captured both if the command is not present compare to default "No such file or director" [1]. If the command is valid but the arguments are incorrect an error is rased giving the command stating that the arguments are wrong. 

@pabloem I have resolved the error in the build concerning the formatting. 

[1] https://stackoverflow.com/questions/24306205/file-not-found-error-when-launching-a-subprocess-containing-piped-commands

**Comment by aaltay** [comment]
@NikeNano According to JIRA if a command fails we only log that such a thing happened but do not log why.

Let's say, we are calling out "pip install a-package-that-does-not-exist", In the error log we would like to see the actual pip error saying that this is an non existent package, in addition to the current error message of a process failed with a non-zero return code. This output will be very likely in stderr.

**Comment by NikeNano** [comment]
Run RAT PreCommi

**Comment by NikeNano** [comment]
Run RAT PreCommit

**Comment by NikeNano** [comment]
@aaltay I have updated and added som tests. The failures atm dosn't seam to be connected to my work but rather something with the jenkins jobs, is that correct?

**Comment by NikeNano** [comment]
@pabloem I have solved all the problems and also added test cases. Would be great if you could provide feedback. 

@aaltay I have updated and added test cases.

**Comment by pabloem** [comment]
Thanks Niklas! Taking a quick look...

**Comment by NikeNano** [comment]
R: @pabloem did you had the opportunity to look at it?

**Comment by pabloem** [comment]
This LGTM. Thanks Niklas! I'll merge now, and if there are more issues, wwe can iterate.

File: sdks/python/apache_beam/utils/processes_test.py [hunk 2]
```diff
@@ -103,5 +104,124 @@ def test_method_forwarding_windows(self, *unused_mocks):
...
+  def setup_class(cls):
+    cls.mock_get_patcher = mock.patch(\
+      'apache_beam.utils.processes.subprocess.check_call')
+    cls.mock_get = cls.mock_get_patcher.start()
+
+  @classmethod
+  def teardown_class(cls):
+    cls.mock_get_patcher.stop()
+
+  def test_oserror_check_call(self):
```
Comment (tvalentyn) [inline sdks/python/apache_beam/utils/processes_test.py]: These tests fail for me when run with 

```
python ./setup.py test -s apache_beam.utils.processes_test.TestErrorHandlingCheckOutput
```
```
======================================================================
ERROR: test_oserror_check_output_message (apache_beam.utils.processes_test.TestErrorHandlingCheckOutput)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/home/projects/beam/beam/beam/sdks/python/apache_beam/utils/processes_test.py", line 162, in test_oserror_check_output_message
    self.mock_get.side_effect = OSError()
AttributeError: 'TestErrorHandlingCheckOutput' object has no attribute 'mock_get'
```
I think the reason is that `setup_class` is never called.
But the tests pass when run inside tox suite:  `tox -e py27`.
Looking further to understand why.

Comment (NikeNano) [reply sdks/python/apache_beam/utils/processes_test.py]: I tested and got the same result, can't understand why though. Did you find anything so far?

Comment (pabloem) [reply sdks/python/apache_beam/utils/processes_test.py]: I might be wrong, but I find that unittest expects the method to be `setUpClass` (https://docs.python.org/3/library/unittest.html#unittest.TestCase.setUpClass). Perhaps tox knows to run `setup_class`, but other way of running the test will miss the `setup_class` method because it's not what unittest expects?

Comment (tvalentyn) [reply sdks/python/apache_beam/utils/processes_test.py]: `setup_class` method is called by nose: https://github.com/nose-devs/nose/blob/7c26ad1e6b7d308cafa328ad34736d34028c122a/nose/suite.py#L134. tox suites executes all tests using nose, however when we run individual tests we run plain `unittest.runner`.

Comment (tvalentyn) [reply sdks/python/apache_beam/utils/processes_test.py]: Looks like `s/setup_class/setUpClass` is a quick fix. I'll send a PR. If I run into additional issues, we can consider rolling this back until a better fix is in place.

Comment (tvalentyn) [reply sdks/python/apache_beam/utils/processes_test.py]: Sent https://github.com/apache/beam/pull/8337.

Comment (NikeNano) [reply sdks/python/apache_beam/utils/processes_test.py]: Sounds good to me.
